### PR TITLE
Issue 674: Render `gui=reverse` color correctly

### PR DESCRIPTION
**Pull request by statiolake**
**Issue #674**
To fix the issue, I added other variants of the related functions which takes a HighlightAttribute and pass the highlight attribute directly instead of reconstructing one. I'd be happy to change the implementation if I should solve this issue in a different way.

### CODE REVIEW

File: src/gui/shellwidget/cell.h [hunk 1]
```diff
@@ -7,6 +7,13 @@
 
 class Cell {
 public:
+	/// Create a cell having a specified HighlightAttribute
+	Cell(uint character, HighlightAttribute const &attribute) noexcept :
```
Comment (jgehrig) [inline src/gui/shellwidget/cell.h:11]: nit: `const HighlightAttribute& attribute`

File: src/gui/shellwidget/shellcontents.h [hunk 2]
```diff
@@ -30,14 +30,7 @@ class ShellContents
...
 		int column,
-		QColor fg = {},
-		QColor bg = {},
-		QColor sp = {},
-		bool bold = false,
-		bool italic = false,
-		bool underline = false,
-		bool undercurl = false,
-		bool reverse = false) noexcept;
+		HighlightAttribute const &hl_attr = HighlightAttribute()) noexcept;
```
Comment (jgehrig) [inline src/gui/shellwidget/shellcontents.h:33]: nit: `const HighlightAttribute& hl_attr = {}`

File: src/gui/shellwidget/shellwidget.h [hunk 3]
```diff
@@ -69,6 +69,12 @@ public slots:
 		bool undercurl = false,
 		bool reverse = false);
 
+	int put(
+		const QString& text,
+		int row,
+		int column,
+		HighlightAttribute const &hl_attr);
```
Comment (jgehrig) [inline src/gui/shellwidget/shellwidget.h:76]: Here also, `const HighlightAttribute& hl_attr`

File: src/gui/shellwidget/highlight.h [hunk 4]
```diff
@@ -25,6 +25,20 @@ class HighlightAttribute {
 	{
 	}
 
+	/// Copies a existing HighlightAttribute.
+	HighlightAttribute(const HighlightAttribute &other) noexcept :
```
Comment (jgehrig) [inline src/gui/shellwidget/highlight.h:29]: We should use the default copy constructor instead. The compiler can automatically generate code like this. http://www.cplusplus.com/articles/y8hv0pDG/

Try simply deleting the code first. If that doesn't work, we may be able to use `HighlightAttribute(const HighlightAttribute&) noexcept = default`? I can never remember the exact rules for implicit CTORS...

**Comment by jgehrig** [comment]
@statiolake

Thanks for the detective work! I wasn't aware of `gui=reverse`, that would have been useful while testing...

The FG/BG flip is a vestige of the old protocol rendering code, and was intentional at the time of writing. I think what you have is a good solution. We should pass raw data to the lower layers and let the renderer handle the flipping. I also think `GetBackground()/GetForeground()` should return the reverse-aware colors.

If I understand correctly, both of these statements are true for your solution.

I ran some tests with the old rendering protocol, `ext_linegrid=false`. Everything looks good with these changes for that protocol too.

File: src/gui/shellwidget/highlight.h (hunk 4 above)
Comment (statiolake) [reply src/gui/shellwidget/highlight.h]: Actually simply deleting those lines worked. I've misunderstood when implicit constructors are defined, and also I've completely forgot about `= default`... They're removed in 7a1434d8a186f16bdf9cf906ab099e584d380430.

**Comment by jgehrig** [comment]
Thanks! The changes look good to me now :)

If you can, it'd be nice to clean up the history with a `git rebase -i`. You should be able to delete `03c8252`/`7a1434d` and squash `c893e25`.

**Comment by jgehrig** [comment]
@equalsraf This looks good to me.

Let's get this merged to resolve Issue 674 and Issue 687.

**Comment by equalsraf** [comment]
ACK, thanks @statiolake

**Pull request merged**